docs/the-new-architecture/cxx-cxxturbomodules.md

```diff
@@ -0,0 +1,522 @@
++ if (name == "NativeSampleModule") {
++   return std::make_shared<facebook::react::NativeSampleModule>(jsInvoker);
++ }
+  return nullptr;
+}
+
+// ...
+```
+
+Update `CMakeLists.txt` with the following entries, e.g., at the very end of that file:
+
+```diff
+// ...
+
+# This file includes all the necessary to let you build your application with the New Architecture.
+include(${REACT_ANDROID_DIR}/cmake-utils/ReactNative-application.cmake)
+
++ # App needs to add and link against tm (TurboModules) folder
++ add_subdirectory(${REACT_ANDROID_DIR}/../../../tm/ tm_build)
++ target_link_libraries(${CMAKE_PROJECT_NAME} tm)
+```
+
+This will instantiante a `NativeSampleModule` associated with the name `NativeSampleModule` as defined in our JavaScript spec file earlier.
+
+## 4. C++ Native Code
+
+For the final step, you'll need to write some native code to connect the JavaScript side to the native platforms. This process requires two main steps:
+
+- Run **Codegen** to see what it generates.
+- Write your native code, implementing the generated interfaces.
+
+### Run code-gen
+
+:::info
+Follow the [Codegen](./pillars-codegen) guide for general information.
+:::
+
+On iOS code-gen is run each time you execute:
+
+```sh
+USE_HERMES=1 RCT_NEW_ARCH_ENABLED=1 bundle exec pod install
+```
+
+You can inspect the generated `AppSpecsJSI.h` and `AppSpecsJSI-generated.cpp` files inside the `CxxTurboModulesGuide/ios/build/generated/ios` folder.
+
+Those files are prefixed with `AppSpecs` as this matches the `codegenConfig.name` parameter added earlier to `package.json`.
+
+On Android code-gen is run each time you execute:
+
+```sh
+yarn android
+```
+
+You can inspect the generated `AppSpecsJSI.h` and `AppSpecsJSI-generated.cpp` files inside the `CxxTurboModulesGuide/android/app/build/generated/source/codegen/jni` folder.
+
+You only need to re-run codegen if you have changed your JavaScript spec.
+
+The C++ function generated for our JavaScript spec file looks like:
+
+```cpp
+virtual jsi::String reverseString(jsi::Runtime &rt, jsi::String input) = 0;
+```
+
+You can directly work with the lower level `jsi::` types - but for convience C++ Turbo Native Modules automatically `bridge` into `std::` types for you.
+
+### Implementation
+
+Now create a `NativeSampleModule.h` file with the following content:
+
+```cpp
+#pragma once
+
+#if __has_include(<React-Codegen/AppSpecsJSI.h>) // CocoaPod headers on Apple
+#include <React-Codegen/AppSpecsJSI.h>
+#elif __has_include("AppSpecsJSI.h") // CMake headers on Android
+#include "AppSpecsJSI.h"
+#else // BUCK headers
+#include <AppSpecs/AppSpecsJSI.h>
+#endif
+#include <memory>
+#include <string>
+
+namespace facebook::react {
+
+class NativeSampleModule : public NativeSampleModuleCxxSpec<NativeSampleModule> {
+ public:
+  NativeSampleModule(std::shared_ptr<CallInvoker> jsInvoker);
+
+  std::string reverseString(jsi::Runtime& rt, std::string input);
+};
+
+} // namespace facebook::react
+```
+
+:::note
```

Comment:
Can we move the Note before the code? `includes` are the first directive that a user reads, so it makes sense to warn them immediately, what do you think?